src/IIIFPresentation/API/Converters/Streaming/StreamingJsonProcessor.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System.Text.Json;
+
+namespace API.Converters.Streaming;
+
+public static class StreamingJsonProcessor
+{
+    /// <summary>
+    ///     Reads <paramref name="input" /> UTF-8 JSON stream, token by token, and writes it to the
+    ///     <paramref name="output" /> also as UTF-8 JSON, optionally performing changes
+    ///     on the fly
+    /// </summary>
+    /// <param name="input">UTF-8 JSON source</param>
+    /// <param name="output">Processed UTF-8 JSON target</param>
+    /// <param name="inputLength">If known, helps to read the stream correctly</param>
+    /// <param name="implementation">Handles actual </param>
+    /// <remarks>
+    ///     In C#13 it can be made async, but currently ref structs don't work with async/await.
+    /// </remarks>
+    public static void ProcessJson(Stream input, Stream output, long? inputLength, IProcessJson implementation)
+    {
+        // Initial buffer size - will auto expand if token/whitespace sequence is bigger than that
+        const int bufferSize = 1024;
```

Comment:
If you can use Newtonsoft.Json instead, the async version is a lot simpler:

```cs
interface JsonProcessor {
    bool PathMatches(string path);

    object? Process(object? value);
}

async void StreamingJsonMap(JsonTextReader reader, JsonTextWriter writer, JsonProcessor processor) {
    while (await reader.ReadAsync()) {
        if (processor.PathMatches(reader.Path) && reader.TokenType != JsonToken.PropertyName) {
            await writer.WriteTokenAsync(reader.TokenType, processor.Process(reader.Value));
        }
        else {
            await writer.WriteTokenAsync(reader.TokenType, reader.Value);
        }
    }
}
```

Reply:
It is simpler, but I ran some benchmarks and it seems the Newtonsoft version takes 5x the time STJ does. As performance was stated to be key factor, the "less simple but much faster" seems to be preferable.